### DISCUSSION TITLE
Can't replace self-signed certificate with signed one. Wildcard certificate and certificateprivatekeypassword.

### DISCUSSION THREAD

**Discussion by Xevichr** [Ylianst/MeshCentral#6178] [Issues] [answered]
Below my config:

```
{
   "settings": {
      "cert": "___", 
      "certificateprivatekeypassword": "___", 
      "_minify": true, 
      "selfupdate": true
   }, 
   "domains": {
      "": {
         "footer": "___"
      }
   }
}
```

Quite a short config. All I would like to do is replace the self-signed certificates generated by MeshCentral with our own. 

1) Its a *.domain.com certificate, a wildcard certificate. Is it possible to use a wildcard certificate with MeshCentral? I can't get it to work. Do I maybe have to change `"cert": "___"` value from `sub.domain.com` to `*.domain.com`? I don't think so, but I just can't get it to work.

2) Besides, that am I even correctly telling MeshCentral what the .key file password is?

Reddit post: https://www.reddit.com/r/MeshCentral/comments/qieoua/troubleshoot_steps_to_update_meshcentral_with/

In this older Reddit post I found the example being `"certificateprivatekeypassword": [ "password1" ]`, but when I use the `[ ]` symbols, after restarting MeshCentral I noticed how it went from `[ "password1" ]` to `["password1"]` in the config file. It removed the spacebars, so I guess not?

Any help would be greatly appreciated as I am closing in on the deadline of my internship project.

**Comment by si458** [comment]
the `cert` value MUST BE A DNS NAME that your remote devices use to communicate with your meshcentral,
it cannot be a wildcard name!
however you can use a wildcard certificate by doing the steps in the reddit post,
and specifying the `certificateprivatekeypassword` if your `.key` file is encrypted

**Reply by Xevichr** [reply]
Ok, so I must use:

`"cert": "sub.domain.com",`

And can enter the private key password with:

`"certificateprivatekeypassword": "___",`

I did just that, but sadly I still couldn't get it to work.

Besides that, in the private key before `-----BEGIN ENCRYPTED PRIVATE KEY-----` there is also some metadata. Can MeshCentral understand this or do I remove it? When I tested it just now I decided to remove it. Previously I tried with.

As mentioned I couldn't get it to work, but also, when applying the new configuration (Under "Edit Configuration" in the MeshCentral Windows Server installation tool), I can't even get the server to restart, so I must be doing something very wrong.

I just replaced the .crt and .key file with my own, without the metadata in front of `-----BEGIN ENCRYPTED PRIVATE KEY-----`, and also have the following config:

```
{
   "settings": {
      "cert": "sub.domain.com", 
      "certificateprivatekeypassword": "___", 
      "_minify": true
   }, 
   "domains": {
      
   }
}

```

I restarted the server with the installation tool by hitting "Edit Configuration" and apply.

**Reply by si458** [reply]
oh your using windows! when you click 'edit' i believe it recreates the config.json file and certificates!
the best way is to STOP the meshcentral service, and then START it using 'windows services'
also logs for the service should be under `c:\program files\open source\meshcentral\winservice\daemon`

**Reply by Xevichr** [reply]
The service shuts itself down after I start it. Through the Windows Services menu as well.

Anyway, the meshcentral.out.log file contains:

```
Installing modules [ 'otplib@10.2.3' ]
MeshCentral HTTP redirection server running on port 80.
Generating certificates, may take a few minutes...
Generating root certificate...
Generating HTTPS certificate...
Generating MeshAgent certificate...
Generating code signing certificate...
Generating Intel AMT MPS certificate...
MeshCentral v1.1.24, Hybrid (LAN + WAN) mode.
Code signed MeshCmd64.exe.
Code signed MeshService64.exe.
Code signed MeshService.exe.
Code signed MeshCmdARM64.exe.
Code signed MeshServiceARM64.exe.
Code signed MeshCmd.exe.
MeshCentral Intel(R) AMT server running on BGP-MESH0001-A.bierensgroup.net:4433.
Server has no users, next new account will be site administrator.
MeshCentral HTTPS server running on BGP-MESH0001-A.bierensgroup.net:443.
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
ERROR: Unable to parse C:\Program Files\Open Source\MeshCentral\meshcentral-data\config.json.
ERROR: Unable to parse C:\Program Files\Open Source\MeshCentral\meshcentral-data\config.json.
ERROR: Unable to parse C:\Program Files\Open Source\MeshCentral\meshcentral-data\config.json.
```

My current config is:

```
{
   "settings": {
      "cert": "sub.domain.com", 
      "certificateprivatekeypassword": [ "___" ],
      "_minify": true, 
      "selfupdate": true
   }, 
   "domains": {
      "": {
         "footer": "___"
      }
   }
}

```

Is it possible anything else went wrong? The password should be correct (for the .key file). In any case, I will look for a way to verify if that is actually the password, just to be sure.

**Reply by si458** [reply]
it thinks your config.json is invalid and missing `certificatePrivateKeyPassword`
`ERROR: Unable to parse C:\Program Files\Open Source\MeshCentral\meshcentral-data\config.json.`
try running the config.json through https://jsonlint.com and it might help find out what is invalid/wrong

p.s: im wondering if your private key has special characters in!? this might be making the json invalid

**Reply by Xevichr** [reply]
So, I ran the jsonlint. It says its valid (both without the [ symbols and with, I am going to assume you're supposed to use it with, considering both the log and the original Reddit post said you're supposed to use it like that).

I don't see any special characters in the Encrypted Private Key (With the blind eye at least).

Thinking about just trying it again without an encrypted private key.

**Reply by si458** [reply]
just checked source code
```
   // Decrypt private key if needed
    obj.decryptPrivateKey = function (key) {
        if (typeof key != 'string') return key;
        var i = key.indexOf('-----BEGIN ENCRYPTED PRIVATE KEY-----');
        var j = key.indexOf('-----END ENCRYPTED PRIVATE KEY-----');
        if ((i >= 0) && (j > i)) {
            var passwords = parent.config.settings.certificateprivatekeypassword;
            if (parent.config.settings.certificateprivatekeypassword == null) { passwords = []; }
            else if (typeof parent.config.settings.certificateprivatekeypassword == 'string') { passwords = [parent.config.settings.certificateprivatekeypassword ]; }
            var privateKey = null;
            for (var k in passwords) { if (privateKey == null) { try { privateKey = obj.pki.decryptRsaPrivateKey(key, passwords[k]); } catch (ex) { } } }
            if (privateKey == null) {
                console.log("Private certificate key is encrypted, but no correct password was found.");
                console.log("Add the password to the \"certificatePrivateKeyPassword\" value in the Settings section of the config.json.");
                console.log("Example: \"certificatePrivateKeyPassword\": [ \"MyPassword\" ]");
                process.exit();
                return null;
            }
            return obj.pki.privateKeyToPem(privateKey);
        }
        return key;
    }
```

so the fact you are getting the message 'Private certificate key is encrypted, but no correct password was found'
means that it tried to decrypt your key file with the password you provided but failed!

one thing you can try is STOP the service, empty the log file, then start the service again, see if the logs are the same or similar

if its the same, then it means the password is incorrect for the key file

make sure they `.key` file is just the BEGIN->KEY->END lines

**Reply by Xevichr** [reply]
I am now getting the following:

```
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
MeshCentral HTTP redirection server running on port 80.
Private certificate key is encrypted, but no correct password was found.
Add the password to the "certificatePrivateKeyPassword" value in the Settings section of the config.json.
Example: "certificatePrivateKeyPassword": [ "MyPassword" ]
```

**Reply by si458** [reply]
ok so the json is valid! (the lines about invalid json might of been old lines)

but again the fact you are getting the message `Private certificate key is encrypted, but no correct password was found`
means that it tried to decrypt your key file with a password you provided but failed!
see comment here - https://github.com/Ylianst/MeshCentral/discussions/6178#discussioncomment-9774330

check you can decrypt the key normally https://marco.maranao.ca/articles/how-decrypt-rsa-private-key-using-openssl

**Reply by Xevichr** [reply, marked as answer]
Alright, so.

I got it to work, this is my eventual config: 

```
{
    "settings": {
        "cert": "sub.domain.com",
        "_minify": true,
        "selfupdate": true
    },
    "domains": {
        "": {
            "footer": "Bierensgroup"
        }
    }
}
```

So the password failed to open the encrypted .key? Likely just a wrong password?

In any case, I got it to work, can't thank you enough this has been a headache hahaha.

**Reply by si458** [reply]
so if you didnt specify a password in your config.json
then im guessing the `.key` file isnt actually encrypted with a password but a blank password!
like how the ssh keys work for remote ssh passwordless logins!

**Reply by Xevichr** [reply]
Well, I did specify the password, it just didn't work I guess.

I received an unencrypted .key file from our partner, that it eventually worked out with just now. 

The partner also provided the encrypted one, with the password for the encrypted .key file (That was definitely encrypted as far as I know because it started with `-----BEGIN ENCRYPTED PRIVATE KEY-----`, and I didn't use `"certificateprivatekeypassword": "___",` in the first place, beacause I didn't know I would have to enter it in the config file, if it wasn't actually encrypted it should've worked then).

Considering  anyone else ever posted something like this regarding the whole `"certificateprivatekeypassword": "___",`, I am going to assume I filled out the password wrong in the config file or received a wrong one in the first place.

**Reply by si458** [reply]
nothing like confusion! at least you got there in the end!
in theory you shouldnt need `certificateprivatekeypassword` as most `.key` are not encrypted with a password

but it would be intresting to know IF the contents of your unencrypted/encrypted key files are the same or different?

and if they are in fact different content, then you should be able to use the encrypted one with `certificateprivatekeypassword`,
and the unencrypted one without `certificateprivatekeypassword`

but i am still wondering if the password they provided you for your encrypted key is actually right!?

**Reply by Xevichr** [reply]
lol, it's too late for me to check. Because certificate .key and .crt files are not really the kind of file you want to have around doing nothing, I deleted them pretty instantly after getting everything to work... I don't have it on me anymore right now, would have to ask them to provide me with one again.
